### PR TITLE
[Custom Ops] Implement TP in linear layers

### LINKED ISSUE #134: Add tp>1 support for Linear Layers
Update SpyreMergedColumnParallelLinear and SpyreRowParallelLinear with weight sharding and communication primitives

### PR DESCRIPTION
**Pull request by R3hankhan123**
## Description
This PR adds support for tensor Parallelism in linear layers. Note once `all_gather` is supported `SpyreMergedColumnParallelLinear` can be removed as its a mirror copy of upstream vllm just adding the runtime warning of all_gather not supported.

all_gather not being supported is not a blocker for most LLMs as gather_output param in model code in vllm have set it to false. This will only effect vision models and MTP models like gemma4 and some qwen 3.5 models
## Related Issues
resolves #134 
## Test Plan
Unit tests have been written and are passing
## Checklist

- [x] I have read the [contributing guidelines](https://torch-spyre.github.io/spyre-inference/contributing/)
- [x] My code follows the project's code style (run `bash format.sh`)
- [x] I have added tests for my changes (if applicable)
- [ ] I have updated the documentation (if applicable)
- [x] My commits include a `Signed-off-by:` line (DCO compliance)

### CODE REVIEW

File: spyre_inference/distributed/spyre_communicator.py [hunk 1]
```diff
@@ -156,21 +156,6 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
         dist.broadcast(input_, src=self.ranks[0], group=self.device_group)
         return input_
 
-    def all_gather(self, input_: torch.Tensor, dim: int = -1) -> torch.Tensor:
```
Comment (R3hankhan123) [inline spyre_inference/distributed/spyre_communicator.py:159]: This is required since its causing error on lm_head logits gathering

File: spyre_inference/custom_ops/linear.py [hunk 2]
```diff
@@ -73,44 +82,130 @@ def process_weights_after_loading(self, layer):
...
+        """
+        bias = self.bias if not self.skip_bias_add else None
+
+        output_parallel = self.quant_method.apply(self, input_, bias)
+
+        if self.gather_output and self.tp_size > 1:
+            raise NotImplementedError(
+                _spyre_collective_unsupported_message(
+                    "allgather",
+                    self.world_size,
```
Comment (joerunde) [inline spyre_inference/custom_ops/linear.py]: self.tp_size here? Not sure self.world_size exists

File: spyre_inference/custom_ops/linear.py [hunk 3]
```diff
@@ -73,44 +82,130 @@ def process_weights_after_loading(self, layer):
...
```
Comment (joerunde) [inline spyre_inference/custom_ops/linear.py]: self.tp_size?

File: spyre_inference/distributed/spyre_communicator.py (hunk 1 above)
Comment (joerunde) [reply spyre_inference/distributed/spyre_communicator.py:159]: Can we get some more info? all_gather isn't implemented yet by spyreccl, and removing this guard here would cause callers to hit a failure further down the stack

File: spyre_inference/custom_ops/linear.py [hunk 4]
```diff
@@ -73,44 +82,130 @@ def process_weights_after_loading(self, layer):
...
         super().__init__(*args, **kwargs)
-        if self.tp_size > 1:
-            raise NotImplementedError(
-                f"{self.__class__.__name__} only supports TP=1, got TP={self.tp_size}"
-            )
 
         if isinstance(self.quant_method, UnquantizedLinearMethod):
             self.quant_method = SpyreUnquantizedLinearMethod()
 
+        logger.debug(
```
Comment (joerunde) [inline spyre_inference/custom_ops/linear.py]: this would probably be too chatty- isn't this gonna log for every linear layer in the model?

File: spyre_inference/custom_ops/linear.py [hunk 5]
```diff
@@ -73,44 +82,130 @@ def process_weights_after_loading(self, layer):
...
+                    self.world_size,
+                    blocker="libspyre_comms list-form allgather + torch-spyre",
+                )
+            )
+
         # D2H before downstream .split() — Spyre can't handle strided views
-        if self.return_bias:
-            return convert(result[0], device="cpu"), result[1]
-        return convert(result, device="cpu")
+        output = convert(output_parallel, device="cpu")
```
Comment (joerunde) [inline spyre_inference/custom_ops/linear.py]: See suggestions here: https://github.com/R3hankhan123/spyre-inference/pull/6

I think we should have a test for when this extra transfer can be removed, and I think we can also use the super() calls for most of these - at least the current tests don't fail so we need more test coverage if there are other required workarounds in these .forward methods

Comment (R3hankhan123) [reply spyre_inference/custom_ops/linear.py]: Thanks @joerunde for the suggestions, I have incorporated the PR

File: spyre_inference/distributed/spyre_communicator.py (hunk 1 above)
Comment (R3hankhan123) [reply spyre_inference/distributed/spyre_communicator.py:159]: https://github.com/vllm-project/vllm/blob/f0204358d9b811bde5320c037236e97b8fb6199d/vllm/platforms/interface.py#L851
ok looks like this is setting it to true

Comment (R3hankhan123) [reply spyre_inference/distributed/spyre_communicator.py:159]: And using this value it determines weather to use all gather or gather
https://github.com/vllm-project/vllm/blob/f0204358d9b811bde5320c037236e97b8fb6199d/vllm/model_executor/layers/logits_processor.py#L75-L87

Comment (R3hankhan123) [reply spyre_inference/distributed/spyre_communicator.py:159]: looks like these logits are on CPU thats why its not failing 
```
(Worker_TP0 pid=121113) INFO 06-03 06:26:19 [logits_processor.py:81] [DEBUG _gather_logits] logits device: cpu, dtype: torch.float16, shape: torch.Size([1, 24608])
(Worker_TP0 pid=121113) INFO 06-03 06:26:19 [logits_processor.py:85] [DEBUG _gather_logits] use_all_gather: True
(Worker_TP0 pid=121113) INFO 06-03 06:26:19 [logits_processor.py:93] [DEBUG _gather_logits] Calling tensor_model_parallel_all_gather()
(Worker_TP0 pid=121113) INFO 06-03 06:26:19 [logits_processor.py:101] [DEBUG _gather_logits] Result device: cpu, shape: torch.Size([1, 49216])
```

Comment (joerunde) [reply spyre_inference/distributed/spyre_communicator.py:159]: Interesting!

It looks like our previous guard here was just too broad- we can check the dtype of the tensors and if they're on cpu we can allow the super() impl to take care of the gather / all_gather. I've reinstated this with that change and it simplifies linear.py a lot too

**Review by joerunde — APPROVED** [review]
(speculative ✅ on changes that are half mine anyway 😉 )

File: spyre_inference/custom_ops/linear.py [hunk 6]
```diff
@@ -73,43 +74,69 @@ def process_weights_after_loading(self, layer):
...
+            "all_gather is not yet supported on Spyre"
+        )
 
     def forward(self, input_):
         result = super().forward(input_)
-        # D2H before downstream .split() — Spyre can't handle strided views
+        # D2H so that GraniteAttention's qkv.split() and the subsequent
+        # v.view() + kv_cache scatter-write run on CPU. Spyre rejects a
+        # non-contiguous tensor as a scatter source; see
+        # test_spyre_strided_scatter_source for the minimal reproduction.
```
Comment (bohnstingl) [inline spyre_inference/custom_ops/linear.py:125]: We may not need to expand this really?

File: spyre_inference/custom_ops/linear.py [hunk 7]
```diff
@@ -73,43 +74,69 @@ def process_weights_after_loading(self, layer):
...
-    `input_` is on `cpu` and from `GraniteMLP` where `input_` is on spyre.
-    Thus, we always convert the `input_` to `spyre`, which is a NoOp in
-    case of `GraniteMLP`.
+    """Spyre RowParallelLinear with TP support.
+
+    Supports TP>=1 with weight sharding along input dimension and
+    all_reduce for aggregating results across ranks when reduce_results=True.
+
+    RowParallelLinear is invoked from GraniteAttention (input on cpu) and
+    GraniteMLP (input on spyre); H2D is a no-op in the latter case.
```
Comment (bohnstingl) [inline spyre_inference/custom_ops/linear.py:139]: I think this comment may not be needed?

File: spyre_inference/distributed/spyre_communicator.py [hunk 8]
```diff
@@ -157,12 +157,17 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
...
         # SpyreCommsContext::allgather (and torch-spyre's spyreccl backend
         # wires up `_allgather_base`), delete this override and let the
         # base class use `dist.all_gather_into_tensor`.
         if self.world_size == 1:
             return input_
+        if input_.device.type == "cpu":
+            return super().all_gather(input_, dim)
         raise NotImplementedError(
             _spyre_collective_unsupported_message(
                 "allgather",
```
Comment (bohnstingl) [inline spyre_inference/distributed/spyre_communicator.py:173]: I am just a little confused by this here. Doesn't tp > 1 on spyre involve `all_gather` and `gather`? From this snippet it seems that it is actually not supported on spyre?

**Review by bohnstingl — CHANGES_REQUESTED** [review]

File: spyre_inference/distributed/spyre_communicator.py (hunk 8 above)
Comment (R3hankhan123) [reply spyre_inference/distributed/spyre_communicator.py:173]: Yes for now `all_gather` and `gather` is not supported for spyre but since logits are on CPU it doesn't fail and `tensor_model_parallel_all_gather` falls back to cpu

Comment (bohnstingl) [reply spyre_inference/distributed/spyre_communicator.py:173]: Okay, so this PR prepares the ground for TP>1 once, we have all_gather and gather really supported on spyre, right?

File: spyre_inference/custom_ops/linear.py (hunk 6 above)
Comment (joerunde) [reply spyre_inference/custom_ops/linear.py:125]: might be too much info, sure. The idea here was though that we could remove this cpu transfer once that test starts passing

File: spyre_inference/distributed/spyre_communicator.py (hunk 8 above)
Comment (joerunde) [reply spyre_inference/distributed/spyre_communicator.py:173]: pretty much!

With the next PR to address lm_head we should at least be able to run TP=2 with the fallback send/recv and the cpu transfer here on SpyreQKVParallelLinear

File: spyre_inference/custom_ops/linear.py (hunk 7 above)
Comment (joerunde) [reply spyre_inference/custom_ops/linear.py:139]: yeah we could remove this entire op probably- just left it here for more context for the reader and future AI agents. I don't have a strong feeling either way

**Review by bohnstingl — APPROVED** [review]
Given the known limitations, fine with me
